learn/generation/langchain/rag-chatbot.ipynb

```diff
@@ -1,1163 +1,2742 @@
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 15,
+      "metadata": {
+        "id": "yg65GnRWZrwL"
+      },
+      "outputs": [],
+      "source": [
+        "from pinecone import Pinecone\n",
+        "\n",
+        "# initialize connection to pinecone (get API key at app.pinecone.io)\n",
+        "api_key = os.getenv(\"PINECONE_API_KEY\") or getpass(\n",
+        "    \"Enter your Pinecone API key: \"\n",
+        ")\n",
+        "\n",
+        "# configure client\n",
+        "pc = Pinecone(api_key=api_key)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "WN6x5Y3vZrwL"
+      },
+      "source": [
+        "Now we setup our index specification, this allows us to define the cloud provider and region where we want to deploy our index. You can find a list of all [available providers and regions here](https://docs.pinecone.io/troubleshooting/available-cloud-regions)."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 16,
+      "metadata": {
+        "id": "T1LKJchlZrwL"
+      },
+      "outputs": [],
+      "source": [
+        "from pinecone import ServerlessSpec\n",
+        "\n",
+        "spec = ServerlessSpec(\n",
+        "    cloud=\"aws\", region=\"us-east-1\"\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "CK13RD1jZrwL"
+      },
+      "source": [
+        "Then we initialize the index. We will be using OpenAI's `text-embedding-ada-002` model for creating the embeddings, so we set the `dimension` to `1536`."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 17,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "I6Iq6WOKZrwL",
+        "outputId": "2af2ccca-0a2e-4204-85d3-aeb0ce04d3f7"
+      },
+      "outputs": [
+        {
+          "data": {
+            "text/plain": [
+              "{'dimension': 1536,\n",
+              " 'index_fullness': 0.0,\n",
+              " 'metric': 'dotproduct',\n",
+              " 'namespaces': {},\n",
+              " 'total_vector_count': 0,\n",
+              " 'vector_type': 'dense'}"
+            ]
+          },
+          "execution_count": 17,
+          "metadata": {},
+          "output_type": "execute_result"
+        }
+      ],
+      "source": [
+        "import time\n",
+        "\n",
+        "index_name = 'deepseek-r1-rag'\n",
+        "existing_indexes = [\n",
+        "    index_info[\"name\"] for index_info in pc.list_indexes()\n",
+        "]\n",
+        "\n",
+        "# check if index already exists (it shouldn't if this is first time)\n",
+        "if index_name not in existing_indexes:\n",
+        "    # if does not exist, create index\n",
+        "    pc.create_index(\n",
+        "        index_name,\n",
+        "        dimension=1536,  # dimensionality of embed 3 small\n",
```

Comment:
I would use the actual model name in this comment

Reply:
updated